// @wazo/sdk/lib/api/accessd.js
'use strict';

require('js-base64');

/*       */

var accessd = (client              , baseUrl        ) => ({
  listSubscriptions: () => client.get(`${baseUrl}/subscriptions`),

  createSubscription: ({ productSku, name, startDate, contractDate, autoRenew, term }        ) => {
    const body = {
      product_sku: productSku,
      name,
      start_date: startDate,
      contract_date: contractDate,
      auto_renew: autoRenew,
      term,
    };

    return client.post(`${baseUrl}/subscriptions`, body);
  },

  getSubscription: (uuid        ) => client.get(`${baseUrl}/subscriptions/${uuid}`),

  deleteSubscription: (uuid        ) => client.delete(`${baseUrl}/subscriptions/${uuid}`),

  createSubscriptionToken: () => client.post(`${baseUrl}/subscriptions/token`),

  listAuthorizations: (subscriptionUuid        ) =>
    client.get(`${baseUrl}/subscriptions/${subscriptionUuid}/authorizations`),

  getAuthorization: (subscriptionUuid        , uuid        ) =>
    client.get(`${baseUrl}/subscriptions/${subscriptionUuid}/authorizations/${uuid}`),

  deleteAuthorization: (subscriptionUuid        , uuid        ) =>
    client.delete(`${baseUrl}/subscriptions/${subscriptionUuid}/authorizations/${uuid}`),

  createAuthorization: (subscriptionUuid        , { startDate, term, service, rules, autoRenew }        ) => {
    const url = `${baseUrl}/subscriptions/${subscriptionUuid}/authorizations`;
    const body = {
      start_date: startDate,
      term,
      service,
      rules,
      auto_renew: autoRenew,
    };

    return client.post(url, body);
  },
});

module.exports = accessd;
