// @wazo/sdk/lib/utils/PhoneNumberUtil.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var LibPhoneNumber = _interopDefault(require('google-libphonenumber'));

//      

const PhoneNumberUtil = LibPhoneNumber.PhoneNumberUtil.getInstance();
const { PhoneNumberFormat, AsYouTypeFormatter } = LibPhoneNumber;

// eslint-disable-next-line
const EXTRA_CHAR_REGEXP = /[^+*\d]/g;

const shouldBeFormatted = (number         ) => {
  if (!number || number.length <= 5) {
    return false;
  }

  return !number.includes('#') && !number.includes('*');
};

const getDisplayableNumber = (number        , country        , asYouType          = false)         => {
  if (!shouldBeFormatted(number)) {
    return number;
  }

  let displayValue = '';

  if (asYouType) {
    const formatter = new AsYouTypeFormatter(country);
    number.split('').forEach(char => {
      displayValue = formatter.inputDigit(char);
    });
  } else {
    const parsedNumber = PhoneNumberUtil.parseAndKeepRawInput(number, country);

    displayValue = PhoneNumberUtil.format(parsedNumber, PhoneNumberFormat.NATIONAL);
  }

  return displayValue;
};

const parsePhoneNumber = (phoneNumber        )         => phoneNumber.replace(EXTRA_CHAR_REGEXP, '');

const getCallableNumber = (number        , country         )          => {
  try {
    if (country) {
      return getDisplayableNumber(number, country).replace(EXTRA_CHAR_REGEXP, '');
    }
    return parsePhoneNumber(number);
  } catch (_) {
    return null;
  }
};

exports.AsYouTypeFormatter = AsYouTypeFormatter;
exports.PhoneNumberFormat = PhoneNumberFormat;
exports.PhoneNumberUtil = PhoneNumberUtil;
exports.getCallableNumber = getCallableNumber;
exports.getDisplayableNumber = getDisplayableNumber;
exports.parsePhoneNumber = parsePhoneNumber;
